### ISSUE TITLE
A "in mail" search in incoming folder will look up YAM when pulling mails

### ISSUE THREAD

**Issue by jens-maus** [jens-maus/yam#220]
**Originally by _mario@cattaneo.de_ on 2010-09-28 16:24:41 +0200**

___
YAM 2.7-dev [OS4/PPC] (25.09.2010) Copyright (C) 2000-2010 YAM Open Source Team

Try to perform a search in "head" or "body" of a mail for "xxx" in "incoming". Now when the search is running (hope you have a folder where it can take a while ;)) push the "get mail" image. Make sure you have new mail waiting to be fetched.
YAM will then lookup/freeze.
I know it's not nice to change the "database" where YAM is searching in, but YAM should never lock up. Should it? ;)

**Comment by tboeckel** [comment]
**Originally on 2010-09-30 09:20:43 +0200**

___
Hm, quite hard to reproduce this issue. So far I was to able to cause any lock ups, no matter how often I searched and downloaded new mails in parallel. The overall progress very much depends on MUI's event handling. Maybe I am doing it in the wrong sequence or in the wrong time frame. Most probably this issue will become more obvious as soon as I have finished the threaded POP3 download.

The filtering works like this: first get a list of all mails in a folder, then perform the search on this (fixed) list of mails. If the download process already downloaded some mails these might end up in this list. If the download process finished while the search is still running and the filtering process moves one of the new mails not yet searched through to a different folder then the search process might stumble over the no longer existing mail.

**Comment by jens-maus** [comment]
**Originally by _mario@cattaneo.de_ on 2010-10-09 17:13:02 +0200**

___
YAM 2.7-dev [OS4/PPC] (09.10.2010) Copyright (C) 2000-2010 YAM Open Source Team

Seems the crash is gone with latest nightly built.

**Comment by jens-maus** [comment]
**Originally by _mario@cattaneo.de_ on 2010-10-09 17:14:55 +0200**

___
It's amazing, 10 tests caused no crash. I resolved this report and *bang*.

Seems Murphy is still with us... :-/

**Comment by tboeckel** [comment]
**Originally on 2010-10-28 13:41:05 +0200**

___
Does this still happen with the recently added thread based POP3 download?

**Comment by jens-maus** [comment]
**Originally by _mario@cattaneo.de_ on 2010-10-28 17:01:53 +0200**

___
Well, with threaded POP3 the download does not start at all.
Clicking the pop-image has no effect, then when I cancel the search the "queued" POP3 is processed.

**Comment by jens-maus** [comment]
**Originally by _mario@cattaneo.de_ on 2010-11-06 15:50:52 +0100**

___
Mind to reply/comment? :)

Is it normal that the buttons does not work anymore, ie. postponed?

**Comment by tboeckel** [comment]
**Originally on 2010-11-08 08:33:53 +0100**

___
Yes, unforturnately this is the normal case. While performing the search YAM does not execute its usual main loop (which handles *all* stuff), but lets itself execute updates of the GUI only to be a bit responsive. But starting a thread requires some communication between the main task and the thread which is not handled during the search. Hence you get the intended reaction right after the search has finished.

It seems we have to move the search to a thread as well...

Setting this ticket back to "accepted" for the time being.

**Comment by tboeckel** [comment]
**Originally on 2011-02-01 11:06:10 +0100**

___
The situation should improve a bit with the next nightly build. That means that YAM will be able again to initiate a download which a search is in progress. But the download will still not be able to download a single mail, because the search process keeps the incoming folder's mail list locked which prevents the download thread from adding a new mail to the same list.

**Comment by jens-maus** [comment]
**Originally on 2011-11-20 23:44:44 +0100**

___
The issue seems to be partly resolved now (accordingly to thore's comments). Thus, I reschedule that ticket to be revised and checked before and eventual 2.8 release so that we can revisit the problem and see if we can get it solved completely by moving things in a separate thread.

**Comment by tboeckel** [comment]
**Originally on 2012-10-11 15:51:07 +0200**

___
In (17ada68):
 * YAM_FI.c: a search within a folder will now clone the folder's list of mails before performing the actual search. This makes it possible to download new mails while the search is still in progress. Please note that the search process may slow down the mail download, because it is not yet done in a separate thread. This finally closes #220.
